iati/validator.py

```diff
@@ -272,6 +272,69 @@ def get_warnings(self):
         return [err for err in self if err.status == 'warning']
 
 
+def _extract_codes_from_attrib(dataset, split_xpath, condition=None):
+    """Extract codes for checking from a Dataset. The codes are being extracted from attributes.
+
+    Args:
+        dataset (iati.data.Dataset): The Dataset to check Codelist values within.
+        split_xpath (list of str): The sections of the XPath, split by `/`.
+        condition (str): An optional XPath expression to limit the scope of what is extracted.
+
+    Returns:
+        list of tuple: A tuple in the format: `(str, int)` - The `str` is a matching code from within the Dataset; The `int` is the sourceline at which the parent element is located.
+
+    """
+    parent_el_xpath = '/'.join(split_xpath[:-1])
+    last_xpath_section = split_xpath[-1:][0]
+
+    attr_name = last_xpath_section[1:]
+    if condition is None:
+        parent_el_xpath = parent_el_xpath + '[@' + attr_name + ']'
+    else:
+        parent_el_xpath = parent_el_xpath + '[' + condition + ' and @' + attr_name + ']'
+
+    # some nasty string manipulation to make the `//@xml:lang` mapping work
+    while not parent_el_xpath.startswith('//'):
+        parent_el_xpath = '/' + parent_el_xpath
+    if parent_el_xpath.startswith('//['):
+        parent_el_xpath = '//*[' + parent_el_xpath[3:]
+
+    parents_to_check = dataset.xml_tree.xpath(parent_el_xpath)
+
+    located_codes = list()
+    for parent in parents_to_check:
+        located_codes.append((parent.attrib[attr_name], parent.sourceline))
+
+    return located_codes
+
+
+def _extract_codes_from_element_text(dataset, split_xpath, condition=None):  # pylint: disable=invalid-name
+    """Extract codes for checking from a Dataset. The codes are being extracted from element text.
+
+    Args:
+        dataset (iati.data.Dataset): The Dataset to check Codelist values within.
+        split_xpath (list of str): The sections of the XPath, split by `/`.
+        condition (str): An optional XPath expression to limit the scope of what is extracted.
+
+    Returns:
+        list of tuple: A tuple in the format: `(str, int)` - The `str` is a matching code from within the Dataset; The `int` is the sourceline at which the parent element is located.
+
+    """
+    parent_el_xpath = '/'.join(split_xpath[:-1])
+
+    # include the condition
+    if condition:
+        parent_el_xpath = parent_el_xpath + '[' + condition + ']'
+
+    parents_to_check = dataset.xml_tree.xpath(parent_el_xpath)
+
+    located_codes = list()
+    for parent in parents_to_check:
+        located_codes.append((parent.text, parent.sourceline))
```

Comment:
I wonder if this should be:

```python
located_codes.append((parent.text.strip(), parent.sourceline))
```

Then the valid test case could be something like:

```xml
<crs-add>
  <channel-code>
    21039
  </channel-code>
</crs-add>
```

Reply:
Matching of Codelists in attributes is exact, with whitespace being deemed incorrect. As such, stripping whitespace for element text doesn't seem like a correct interpretation. Maybe a warning?

Reply:
So I agree for attributes… I’d be tempted to be more forgiving for element text. As a quite hand-wavy justification, IATI-Rulesets allows for whitespace around an `iati-identifier`:

 * https://github.com/IATI/IATI-Rulesets/blob/c33d11885e1084bb99d897c9a5dc3bfcea91c0fb/iatirulesets/__init__.py#L63
 * https://github.com/IATI/IATI-Rulesets/blob/c33d11885e1084bb99d897c9a5dc3bfcea91c0fb/rulesets/standard.json#L19-L20

…so the following passes:

```xml
<iati-identifier>
  some-valid-identifier
</iati-identifier>
```

Reply:
> As a quite hand-wavy justification, IATI-Rulesets allows for whitespace around an iati-identifier

In that same implementation, [startswith](https://github.com/IATI/IATI-Rulesets/blob/c33d11885e1084bb99d897c9a5dc3bfcea91c0fb/iatirulesets/__init__.py#L74) [doesn't permit whitespace](https://github.com/IATI/IATI-Rulesets/blob/c33d11885e1084bb99d897c9a5dc3bfcea91c0fb/iatirulesets/__init__.py#L11) leading whitespace.

---

The source to go back to is the [XPath spec](https://www.w3.org/TR/1999/REC-xpath-19991116/#node-tests) which shows that `text()` returns the entire [text node](https://www.w3.org/TR/1999/REC-xpath-19991116/#section-Text-Nodes). The XPath in the mapping file says that the comparison is against this value (rather than a trimmed or stripped version). As such, checking against a modified text node is not correct.

Further, the [XML Spec](https://www.w3.org/TR/xml/#sec-white-space) shows that white space is significant and must be treated as part of a value - treating it as not-present is incorrect.

Reply:
> In that same implementation, startswith doesn't permit whitespace leading whitespace.

Well, that’s not used in the standard ruleset… But yes – I take your point and I do agree!

It looks like slovakaid are the only publisher currently using the element anyway, and [they get it right](https://iatiregistry.org/dataset/slovakaid-69_1_ac). Also, very few publishers seem to make the mistake of whitespace around `iati-identifier`s.